python/hail/dataset.py

```diff
@@ -349,6 +359,52 @@ def annotate_alleles_expr(self, expr, propagate_gq=False):
         return VariantDataset(self.hc, jvds)
 
     @handle_py4j
+    def annotate_genotypes_expr(self, expr):
+        """Annotate genotypes with expression.
+
+        **Examples**
+
+        Convert the genotype schema to a :py:class:`~hail.expr.TStruct` with two fields ``GT`` and ``CASE_HET``:
+
+        >>> vds_result = vds.annotate_genotypes_expr('g = {GT: g.gt, CASE_HET: sa.pheno.isCase && g.isHet()}')
+
+        Assume a VCF is imported with ``generic=True`` and the resulting genotype schema
+        is a ``Struct`` and the field ``GTA`` is a ``Call`` type. Use the ``Genotype()`` constructor in the
+        expression language to convert a ``Call`` to a ``Genotype``. ``vds_gta`` will have a genotype schema equal to
+        :py:class:`~hail.expr.TGenotype`
+
+        >>> vds_gta = (hc.import_vcf('data/example3.vcf.bgz', generic=True, call_fields=['GTA'])
+        ...                 .annotate_genotypes_expr('g = Genotype(g.GTA)'))
+
+        **Notes**
+
+        .. warning::
+
+            - If the resulting genotype schema is not :py:class:`~hail.expr.TGenotype`,
+            subsequent function calls on the annotated variant dataset may not work such as
+            :py:meth:`~hail.VariantDataset.pca` and :py:meth:`~hail.VariantDataset.linreg`.
+
+            - Hail performance may be significantly slower (~2X) if the annotated variant dataset does not have a
+            genotype schema equal to :py:class:`~hail.expr.TGenotype`.
+
```

Comment:
add a warning saying that genotypes are immutable, and that `g.gt = g.gt + 1` returns a struct rather than modifying the genotype